store/models/log_events.go

```diff
@@ -274,38 +269,89 @@ func (le RunLogEvent) Requester() common.Address {
 	return common.BytesToAddress(b)
 }
 
+func (le RunLogEvent) Payment() common.Address {
+	b := le.Log.Topics[RequestLogTopicRequester].Bytes()
+	return common.BytesToAddress(b)
+}
+
 // JSON decodes the CBOR in the ABI of the log event.
 func (le RunLogEvent) JSON() (JSON, error) {
-	el := le.Log
-	js, err := decodeABIToJSON(el.Data)
+	//el := le.Log
+	//js, err := decodeABIToJSON(el.Data)
+	//if err != nil {
+	//return js, err
+	//}
+
+	topic, err := le.getTopic(0)
+	if err != nil {
+		return JSON{}, err
+	}
+	parse, ok := topicFactoryMap[topic]
+	if !ok {
+		return JSON{}, fmt.Errorf("No parser for the RunLogEvent topic %v", topic)
+	}
+	return parse(le)
+}
+
+var evmWordSize = common.HashLength
+var idSize = evmWordSize
+var versionSize = evmWordSize
+var callbackAddrSize = evmWordSize
+var callbackFuncSize = evmWordSize
+var expirationSize = evmWordSize
+var dataLocationSize = evmWordSize
+var dataLengthSize = evmWordSize
+
+func parseRunLog(le RunLogEvent) (JSON, error) {
+	data := le.Log.Data
+	start := idSize + versionSize + dataLocationSize + dataLengthSize
+
+	js, err := ParseCBOR(data[start:])
+	if err != nil {
+		return js, err
+	}
+	js, err = js.Add("address", le.Log.Address.String())
 	if err != nil {
 		return js, err
 	}
 
-	fullfillmentJSON, err := fulfillmentToJSON(el)
+	js, err = js.Add("dataPrefix", bytesToHex(data[:idSize]))
 	if err != nil {
 		return js, err
 	}
-	return js.Merge(fullfillmentJSON)
+
+	return js.Add("functionSelector", OracleFulfillmentFunctionID)
 }
 
-func fulfillmentToJSON(el Log) (JSON, error) {
-	var js JSON
-	js, err := js.Add("address", el.Address.String())
+func parseRunLog2(le RunLogEvent) (JSON, error) {
+	data := le.Log.Data
```

Comment:
How do you propose testing these if they are private? And would this be more testable if it just took a topic and `Data`?

Reply:
I was just planning on testing them through `RunLogEvent.JSON`, that seemed like the entry point for testing before and I thought it would be sufficient, but the table could easily grow massive.

It's possibly more testable if it accepts a topic, but Data isn't enough. We need the RunLogEvent to pull out the payment amount from the topics.

Reply:
In general, it's best to try and get as reasonably close to the method you're testing to reduce bootstrap and setup test code. By relying on `RunLogEvent` you force the creation of that instance, which depends on an `Initiator`, etc. Furthermore, you kind of have a dependency loop:

```golang
func (le RunLogEvent) JSON() (JSON, error)

// which then calls parseRunLog*, passing itself

func parseRunLog20190123(le RunLogEvent) (JSON, error)
```

After looking at this method, I think it would be best if it just took the `Log`:

`ParseRunLog20190123(log models.Log) (JSON, error)`

I guess at that point, you might be tempted to just have `RunLogEvent#JSON()` just `return parse(le.Log)` and have it do the topic lookup. That's not a bad idea:

```golang
func (le RunLogEvent) JSON() (JSON, error) {
    return ParseRunLog(le.Log)
}
```

I think this is a clean hand off of responsibility, and should ease testing. You can also see for future maintenance, this could easily be put into its own package or type.

Reply:
Pulled out the ParseRunLog function, and replaced the JSON tests with that. Definitely simpler now.